lib/experimental/full-page-client-side-navigation.php

```diff
@@ -0,0 +1,60 @@
+<?php
+/**
+ * Registers full page client-side navigation option using the Interactivity API and adds the necessary directives.
+ */
+
+/**
+ * Enqueue the interactivity router script.
+ */
+function _gutenberg_enqueue_interactivity_router() {
+	// Set the navigation mode to full page client-side navigation.
+	wp_interactivity_config( 'core/router', array( 'navigationMode' => 'fullPage' ) );
+	wp_enqueue_script_module( '@wordpress/interactivity-router' );
+}
+
+add_action( 'wp_enqueue_scripts', '_gutenberg_enqueue_interactivity_router' );
+
+/**
+ * Set enhancedPagination attribute for query loop when the experiment is enabled.
+ *
+ * @param array $parsed_block The parsed block.
+ *
+ * @return array The same parsed block with the modified attribute.
+ */
+function _gutenberg_add_enhanced_pagination_to_query_block( $parsed_block ) {
+	if ( 'core/query' !== $parsed_block['blockName'] ) {
+		return $parsed_block;
+	}
+
+	$parsed_block['attrs']['enhancedPagination'] = true;
+	return $parsed_block;
+}
+
+add_filter( 'render_block_data', '_gutenberg_add_enhanced_pagination_to_query_block' );
+
+/**
+ * Add directives to all links.
+ *
+ * Note: This should probably be done per site, not by default when this option is enabled.
+ *
+ * @param array $content The block content.
+ *
+ * @return array The same block content with the directives needed.
+ */
+function _gutenberg_add_client_side_navigation_directives( $content ) {
+	$p = new WP_HTML_Tag_Processor( $content );
+	while ( $p->next_tag( array( 'tag_name' => 'a' ) ) ) {
+		if ( empty( $p->get_attribute( 'data-wp-on--click' ) ) ) {
+			$p->set_attribute( 'data-wp-on--click', 'core/router::actions.navigate' );
+		}
+		if ( empty( $p->get_attribute( 'data-wp-on--mouseenter' ) ) ) {
+			$p->set_attribute( 'data-wp-on--mouseenter', 'core/router::actions.prefetch' );
+		}
+	}
+	// Hack to add the necessary directives to the body tag.
+	// TODO: Find a proper way to add directives to the body tag.
+	return (string) $p . '<body data-wp-interactive="core/experimental" data-wp-context="{}">';
```

Comment:
We could use a static variable here to add the `<body>` tag once.

```suggestion
	static $body_interactive_added;
	if ( ! $body_interactive_added ) {
		$body_interactive_added = true;
		return (string) $p . '<body data-wp-interactive="core/experimental" data-wp-context="{}">';
	}

	return (string) $p;
```

Reply:
That'd be important to ensure it gets added only once.

Reply:
Is this injecting an additional `<body>` tag after the `body` has already been opened? Like this?

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
  </head>
  <body>
    <h1>
      Hello World!
    </h1>
    <body class="hey">
  </body>
</html>
```

Surprisingly, this actually works:

![image](https://github.com/WordPress/gutenberg/assets/134745/5337e3fb-1672-4683-af7a-ba9d63e69501)

Maybe I shouldn't be surprised given HTML's loose parsing rules. But I can imagine other plugins that try to parse the HTML to, for example, apply various optimizations would get might confused when encountering multiple `<body>` tags.

Shouldn't the attributes rather get injected on the existing `body` tag instead with the HTML Tag Processor? I see this is not currently possible since the `body` tag is hard-coded in [`template-canvas.php`](https://github.com/WordPress/wordpress-develop/blob/f525e665b6ea6e88289d82cad5fc1dfac57db109/src/wp-includes/template-canvas.php#L20). This brings up the question again of output buffering for the entire template (Core-43258). It's something I've been working on in the context of the Performance team's [Optimization Detective](https://wordpress.org/plugins/optimization-detective/) plugin which output buffers the entire template and then uses HTML Tag Processor to do optimizations.

Reply:
I've added the check in [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/57ef83ceb0b51a09e522d5bea4218c5acb755c1d) and I added a follow-up task in [this list](https://github.com/WordPress/gutenberg/pull/59707#issuecomment-2058832203) to explore how to do this properly.

Reply:
Opened PR to address this: https://github.com/WordPress/gutenberg/pull/61212

Reply:
Nice, thank you for sharing @westonruter.